Flax pipeline pndm

pcuenca opened the pull request:
WIP for PNDM scheduler support in Flax pipeline. Includes #577 and #581 (which I need for testing).

patil-suraj commented on src/diffusers/configuration_utils.py:
```diff
@@ -154,15 +154,25 @@ def from_config(cls, pretrained_model_name_or_path: Union[str, os.PathLike], ret
         model = cls(**init_dict)
+        return_tuple = (model,)
+
+        # Some components (Flax schedulers) have a state.
+        if getattr(cls, "has_state", False):       # Check for "create_state" in model instead?
+            state = model.create_state()
+            return_tuple += (state,)
```
We should make this a bit more robust

1. Check if it's a scheduler i.e instance of `SchedulerMixin` if yes
2. Check if the name starts with `Flax`

patil-suraj commented on src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:
```diff
@@ -172,7 +172,7 @@ def __call__(
         )
         if latents is None:
-            latents = jax.random.normal(prng_seed, shape=latents_shape, dtype=self.dtype)
+            latents = jax.random.normal(prng_seed, shape=latents_shape, dtype=jnp.float32)
```
can we use the models `dtype` here ?

pcuenca replied:
`self.dtype` didn't work for me when using `bfloat16`. The latents generated in the denoising loop are always computed in `float32`, so they don't match the previous type if it was something else.

patil-suraj commented on src/diffusers/schedulers/scheduling_ddim_flax.py:
```diff
@@ -104,6 +104,7 @@ class FlaxDDIMScheduler(SchedulerMixin, ConfigMixin):
             `set_alpha_to_one=False`, to make the last step use step 0 for the previous alpha product, as done in
             stable diffusion.
     """
+    has_state = True
```
Think this should be a  read-only `property` instead of class variable so it can't be modified by users.

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -112,7 +113,8 @@ class FlaxPNDMScheduler(SchedulerMixin, ConfigMixin):
             stable diffusion.
     """
-
+    has_state = True
+    
```
same comment as above

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -223,73 +233,77 @@ def step(
-
-        if not return_dict:
-            return (prev_sample, state)
+        return self.step_plms(
+            state=state, model_output=model_output, timestep=timestep, sample=sample, return_dict=return_dict
+        )
```
what's the issue with `prk` step ?

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -330,29 +344,85 @@ def step_plms(
             )
 
+        def counter_2(state: PNDMSchedulerState):
+            ets = state.ets.at[1].set(model_output)
+            return state.replace(
+                ets = ets,
+                cur_model_output = (3 * ets[1] - ets[0]) / 2,
+                cur_sample = sample,
+            )
+
+        def counter_3(state: PNDMSchedulerState):
+            ets = state.ets.at[2].set(model_output)
+            return state.replace(
+                ets = ets,
+                cur_model_output = (23 * ets[2] - 16 * ets[1] + 5 * ets[0]) / 12,
+                cur_sample = sample,
+            )            
+
+        def counter_other(state: PNDMSchedulerState):
+            ets = state.ets.at[3].set(model_output)
+            next_model_output = (1 / 24) * (55 * ets[3] - 59 * ets[2] + 37 * ets[1] - 9 * ets[0])
+
+            ets = ets.at[0].set(ets[1])
+            ets = ets.at[1].set(ets[2])
+            ets = ets.at[2].set(ets[3])
+
```
Think if/else should also work here. Did not work ? what was the issue ?

pcuenca replied:
Resolved offline after reviewing https://github.com/patil-suraj/stable-diffusion-jax/pull/5 and https://github.com/patil-suraj/stable-diffusion-jax/pull/8.

pcuenca commented:
This is ready for review, I just synced with `main`. There are almost more commits than line changes lol, because I kept changes from other PRs that I needed to test.

I tested a generation with the current Flax pipeline using the `FlaxPNDMScheduler` and it worked fine, so we may want to set it as the default for Stable Diffusion after we verify everything's correct. In that regard, I tested the `plms` steps extensively when I was working in Suraj's repo, and I haven't tested the `prk` steps at all – just checked that they run. Perhaps we can use @kashif's tests to verify.

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -196,8 +197,11 @@ def set_timesteps(self, state: PNDMSchedulerState, num_inference_steps: int) ->
             timesteps=jnp.concatenate([state.prk_timesteps, state.plms_timesteps]).astype(jnp.int64),
-            ets=jnp.array([]),
             counter=0,
+            # Will be zeros, not really empty
```
(nit) then why not use `zeros`

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -227,22 +231,32 @@ def step(
-                state=state, model_output=model_output, timestep=timestep, sample=sample, return_dict=return_dict
+            prev_sample, state = jax.lax.switch(
+                jnp.where(state.counter < len(state.prk_timesteps), 0, 1),
+                [self.step_prk, self.step_plms],
```
(nit)
```suggestion
                (self.step_prk, self.step_plms),
```

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -266,42 +280,70 @@ def step_prk(
+
+        state, model_output = jax.lax.switch(
+            state.counter % 4,
+            [remainder_0, remainder_1, remainder_2, remainder_3],
```
(nit)
```suggestion
            (remainder_0, remainder_1, remainder_2, remainder_3),
```

patil-suraj commented on src/diffusers/schedulers/scheduling_pndm_flax.py:
```diff
@@ -266,42 +280,70 @@ def step_prk(
+        )
+
+        # if state.counter % 4 == 0:
+        #     state = state.replace(
+        #         cur_model_output=state.cur_model_output + 1 / 6 * model_output,
+        #         ets=state.ets.append(model_output),
+        #         cur_sample=sample,
+        #     )
+        # elif (self.counter - 1) % 4 == 0:
+        #     state = state.replace(cur_model_output=state.cur_model_output + 1 / 3 * model_output)
+        # elif (self.counter - 2) % 4 == 0:
+        #     state = state.replace(cur_model_output=state.cur_model_output + 1 / 3 * model_output)
+        # elif (self.counter - 3) % 4 == 0:
+        #     model_output = state.cur_model_output + 1 / 6 * model_output
+        #     state = state.replace(cur_model_output=0)
 
         # cur_sample should not be `None`
-        cur_sample = state.cur_sample if state.cur_sample is not None else sample
+        # cur_sample = state.cur_sample if state.cur_sample is not None else sample
```
should be deleted before merging

patil-suraj reviewed (approved):
Very cool @pcuenca , just left tiny nits.

And yes, we can set this as default scheduler, maybe in a follow-up PR.

Thanks a lot for working on this :)

patrickvonplaten commented:
Should we merge this one?

pcuenca commented:
> Should we merge this one?

I think I addressed all the comments, so it should be ok to merge.

We still need to verify the numerical outputs of the PRK steps (not used in Stable Diffusion afaik). I'll try to do it today.

patrickvonplaten commented:
Merging then! Think we don't need to verify prk as it's Flax. Let's only dive into it if people ask for it
